### PR TITLE
`container_app_job`: support secrets from Key Vault

### LINKED ISSUE #25934: `azurerm_container_app_job` can't create secret that references KeyValt
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.5.3

### AzureRM Provider Version

3.103.1

### Affected Resource(s)/Data Source(s)

azurerm_container_app_job

### Terraform Configuration Files

```hcl
data "azurerm_key_vault_secret" "api_key" {
  name         = "api-key"
  key_vault_id = azurerm_key_vault.default.id
  depends_on = [azurerm_role_assignment.key_vault_administrators_key_vault_secrets_officer]
}

resource "azurerm_container_app_job" "job" {
  name                         = "job"
  container_app_environment_id = azurerm_container_app_environment.default.id
  resource_group_name          = azurerm_resource_group.default.name
  location                     = azurerm_resource_group.default.location
  replica_timeout_in_seconds = 60
  replica_retry_limit        = 1

  identity {
    type         = "UserAssigned"
    identity_ids = [azurerm_user_assigned_identity.default.id]
  }

  secrets {
    name = "api-key"
    key_vault_secret_id = data.azurerm_key_vault_secret.api_key.id
    identity = azurerm_user_assigned_identity.default.id
  }

  template {
    container {
      name   = "worker"
      image  = ""
      cpu    = 0.5
      memory = "1Gi"

      env {
        name = "PLANG_API_KEY"
        secret_name = "plang-api-key"
      }
    }
  }
}
```

### Debug Output/Panic Output

```shell
│ Error: updating Job (Subscription: "002c35be-da1d-4a33-a243-9c11b7aaf0cd"
│ Resource Group Name: "resource-group"
│ Job Name: "job"): performing CreateOrUpdate: unexpected status 400 (400 Bad Request) with error: ContainerAppSecretInvalid: Invalid Request: Container app secret(s) with name(s) 'api-key' are invalid: value or keyVaultUrl and identity should be provided.
│ 
│   with module.mymodule.azurerm_container_app_job.job,
│   on mymodule/container_app.tf line 8, in resource "azurerm_container_app_job" "job":
│    8: resource "azurerm_container_app_job" "job" {
│ 
│ updating Job (Subscription: ""
│ Resource Group Name: "resourcegroup"
│ Job Name: "job"): performing CreateOrUpdate: unexpected status 400 (400 Bad Request) with error:
│ ContainerAppSecretInvalid: Invalid Request: Container app secret(s) with name(s) 'api-key' are invalid: value or
│ keyVaultUrl and identity should be provided.
```

### Expected Behaviour

- This should be accepted as `key_vault_secret_id` and `identity` is provided
- This works for `azurerm_container_app`

### Actual Behaviour

A 400 bad request error is returned (see above)

### Steps to Reproduce

1) Add a `secrets` block that references a key vault secret to a `azurerm_container_app_job` resource

For example:
```hcl
resource "azurerm_container_app_job" "job" {
  ...

  secrets {
    name = "api-key"
    key_vault_secret_id = data.azurerm_key_vault_secret.api_key.id
    identity = azurerm_user_assigned_identity.default.id
  }

  ...
}
```

2) Run `terraform apply`

### Important Factoids

_No response_

### References

_No response_

### LINKED ISSUE #25994: azurerm_container_app_job - 'cannot remove secrets from Container Apps at this time due to a limitation in the Container Apps Service'
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

When deploying a container app job, you get the following error when trying to remove a secret:

`│ Error: cannot remove secrets from Container Apps at this time due to a limitation in the Container Apps Service. Please see https://github.com/microsoft/azure-container-apps/issues/395 for more details`

This was changed/resolved for the container app resource #24945 I am hoping the same can occur for container app jobs.

Thank you

### New or Affected Resource(s)/Data Source(s)

azurem_container_app_job

### Potential Terraform Configuration

_No response_

### References

**Related issue/PR**
https://github.com/hashicorp/terraform-provider-azurerm/issues/24945
https://github.com/hashicorp/terraform-provider-azurerm/pull/25743

### PR DESCRIPTION
**Pull request by Botje**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR adds support for referencing Key Vault secrets from Container App Jobs and lifts the restriction on removing secrets.
This code is ported from the Container Apps resource.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<details><summary>Testing evidence</summary>
<pre>
TF_ACC=1 go test -v ./internal/services/containerapps -run=TestAccContainerAppJob_withKeyVault -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccContainerAppJob_withKeyVaultSecret
=== PAUSE TestAccContainerAppJob_withKeyVaultSecret
=== RUN   TestAccContainerAppJob_withKeyVaultSecretVersioningUpdate
=== PAUSE TestAccContainerAppJob_withKeyVaultSecretVersioningUpdate
=== RUN   TestAccContainerAppJob_withKeyVaultSecretIdentityUpdate
=== PAUSE TestAccContainerAppJob_withKeyVaultSecretIdentityUpdate
=== CONT  TestAccContainerAppJob_withKeyVaultSecret
=== CONT  TestAccContainerAppJob_withKeyVaultSecretIdentityUpdate
=== CONT  TestAccContainerAppJob_withKeyVaultSecretVersioningUpdate
--- PASS: TestAccContainerAppJob_withKeyVaultSecret (894.40s)
--- PASS: TestAccContainerAppJob_withKeyVaultSecretVersioningUpdate (974.74s)
--- PASS: TestAccContainerAppJob_withKeyVaultSecretIdentityUpdate (1019.78s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/containerapps	1022.386s
</pre>
</details>

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `container_app_job` - support secrets from Key Vault [GH-0000]
* `container_app_job` - allow removing secrets [GH-0000]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #25934
Fixes #25994 

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @Botje - We're having trouble running the tests for this PR due to some dependency changes/issues with `main` that have crept in somewhere. Are you able to update your fork's main and rebase to allow us to run the tests?

**Comment by Botje** [comment]
Despite subscribing to the linked issue, I did not get a notification. I will update the tests anyway, as promised.

**Comment by Botje** [comment]
Rebased on current `main` branch and converted the tests to use `secret`.

File: website/docs/r/container_app_job.html.markdown [hunk 1]
```diff
@@ -314,9 +314,19 @@ A `volume` block supports the following:
...
+* `identity` - (Optional) The identity to use for accessing the Key Vault secret reference. This can either be the Resource ID of a User Assigned Identity, or `System` for the System Assigned Identity.
 
-* `value` - (required) Value of the secret.
+!> **Note:** `identity` must be used together with `key_vault_secret_id`
+
+* `key_vault_secret_id` - (Optional) The ID of a Key Vault secret. This can be a versioned or version-less ID.
+
+!> **Note:** When using `key_vault_secret_id`, `ignore_changes` should be used to ignore any changes to `value`.
+
+* `name` - (Required) The secret name.
```
Comment (jackofallops) [inline website/docs/r/container_app_job.html.markdown:325]: Can we have this back at the top of the block as it's a required property? (this takes precedence over alphabetical ordering)

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @Botje - Thanks for the rebase and update. Just one minor docs change to look at and then I think this will be good to go, pending testing.

File: website/docs/r/container_app_job.html.markdown (hunk 1 above)
Comment (Botje) [reply website/docs/r/container_app_job.html.markdown:325]: Sure thing. I got this from `container_app` so I will reorder it there as well, then.

**Review by jackofallops — APPROVED** [review]
Thanks @Botje for the updates.  This LGTM, I'll get the tests run now.

**Comment by jackofallops** [comment]
Tests look fine (failures are unrelated to change)

image

**Pull request merged**